Reconnect Notice: In-place Reconnect

sergeymitr opened the pull request:
#### Changes proposed in this Pull Request:
This PR introduces a reconnection flow, which allows users to reconnect Jetpack in-place, in case of connection errors.
In particular, this PR:
- Adds a `connection/reconnect` endpoint, that accepts `POST` requests, with a **required** `action` parameter. The `action` parameter for now can be only `reconnect`. This corresponds to re-registering the site and deleting all tokens (user and blog), so the users need to re-authorize wpcom and re-issue their blog tokens. In the future, this endpoint will also support the actions: `refresh_blog_token` and `refresh_user_token`. 
- Displays an error notice, in cases of connection errors, along with a `Restore connection` action prompt, that triggers the reconnection flow
- Displays the authorization iframe so that a user can re-connect their wpcom account in-place, as part of the reconnection flow (after successful response from the `connection/reconnect` endpoint the user needs to re-connect because the corresponding user tokens are now deleted). By setting the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant  `true` or using **Safari** browser, instead of displaying the authorization iframe the user is redirected to wpcom in order to authorize their account.
- Moves all related Admin page files to using Core translation functions in the context of [Admin Page: Move from i18n-calypso to Core functions to handle translations](https://github.com/Automattic/jetpack/issues/16481)

#### Jetpack product discussion
`p9dueE-1CF-p2`

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

**Prerequisites:**
- A "broken" Jetpack connection: Using the Broken Token plugin, invalidate the **blog token**.

**Testing:**
- Make sure after "breaking" the blog token, the connection error is visible along with the `Restore connection` CTA.
- Click the `Restore connection` CTA: The admin UI should blur and you should see an info message that reconnection is in progress
- The authorization iframe should be displayed in order to re-connect your wpcom account
- After re-connecting your wpcom account the page will refresh and your connection should be now healthy (the connection error shouldn't display anymore)

**Note:** By setting the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant  `true` or testing this with **Safari** browser, instead of seeing the authorization iframe you should be redirected to wpcom in order to authorize your account.

**Before:**
N/A

**After:**
https://d.pr/v/oN8gUa

#### Proposed changelog entry for your changes:
* Easy reconnect in case of connection errors.

sergeymitr committed:
Reconnect Notice: Making the "Reconnect" button initiate the "in-place" connection process.

fgiannar committed:
Rest api: Add reconnect endpoint

fgiannar committed:
Connection state: Add reconnect site actions

fgiannar committed:
Connection state: Update reconnectSite

fgiannar committed:
Jetpack notices: Tidy up connection error notices

fgiannar committed:
Connection state: When auth in place while reconnecting mark successful reconnection after auth

fgiannar committed:
Reconnet flow: Blur main content while reconnecting

sergeymitr committed:
Reconnect Notice: Minor refactoring of the supported actions rendering.

Previously written "supported actions" code structure no longer responds to current needs and leads to code repetition.
I cleaned this up a little bit to simplify the component.
The functionality did not change.

sergeymitr committed:
Adding new `connection/reconnect` REST API endpoint.

The endpoint disconnects the site, initializes the new connection, and returns the authorization URL for Calypso or in-place flow.

fgiannar committed:
AuthIframe: Remove scroll effect on reconnection flow

fgiannar committed:
Connection state: Update success notice on reconnection flow

fgiannar committed:
Jetpack action reconnect notice: Disable click if site is already reconnecting

fgiannar committed:
Jetpack action reconnect notice: Update legend

fgiannar committed:
AuthIframe: Display iframe title

fgiannar committed:
Reconnection flow: Reload current page on successful reconnection

jeherve commented:
As you work on this, do you think you could start using `@wordpress/i18n` for all new translatable strings instead of the existing `i18n-calypso`? This should save us a bit of rebase troubles as we make progress on #16481.

sergeymitr committed:
Reconnect notice: Show error message or code on failure.

If reconnecting fails, we display an error. This commit makes that logic work more smoothly:
1. Instead of the `{ status: 'failure' }`, we return an actual error 500 with the error code and message (if exists) in the body.
2. The error message or code is dispatched further, so the actual error is displayed in the "Failure" badge.

fgiannar committed:
Connection state: Do not dispatch successful auth on reconnection flow

fgiannar committed:
Connection banner: Hide on connection and site data errors

fgiannar committed:
Admin Page: Move to Core translation functions

fgiannar commented on _inc/client/state/connection/actions.js:
```diff
@@ -309,11 +315,11 @@ export const reconnectSite = ( action = 'reconnect' ) => {
+						sprintf(
+							/* translators: placeholder is the error. */
+							__( 'There was an error reconnecting Jetpack. Error: %s', 'jetpack' ),
+							error.response.message || error.response.code
```
@sergeymitr @jeherve fyi this was accidentally [reverted in a previous commit ](https://github.com/Automattic/jetpack/pull/16475/commits/bbd297688a0ba4eb44e692ae71dd9bacfb0662d8)and took the opportunity to fix it here

sergeymitr replied:
Good catch, thanks 👍

sergeymitr committed:
Reconnect notice: Blur the dashboard immediately after reconnect is requested.

Although the `isReconnectSite` state property is set immediately after the "Restore Connection" button is clicked, the `Main` component didn't allow re-rendering.
This commit makes `Main` component to check if `isReconnectSite` has changed, and allows re-rendering if it has, so the dashboard gets blurred.

sergeymitr committed:
Reconnect process: Removing unused code.

sergeymitr committed:
Reconnect process: Unit testing for the `connection/reconnect` endpoint.

- Unit test for the endpoint.
- The new test made `ManagerTest::test_api_url_uses_constants_and_filters()` fail, so I had to adjust that too to make it work.

kbrown9 reviewed (changes requested):
This works well for me. I tested with the iframe and Calypso flow. I just have a couple of comments about the code.

sergeymitr committed:
Reconnect Notice: `connection/reconnection` endpoint capability verification.

- The endpoint `connection/reconnect` now requires the capability `jetpack_disconnect` to be assigned to the active user.
- `JetpackConnectionErrors` doesn't rely on the Redux state, so we get rid of the `connect()` call for that component.
- `JetpackConnectionErrors` had `key` missing for the `React.Fragment` blocks, fixed now.

sergeymitr commented on _inc/client/components/jetpack-notices/jetpack-connection-errors.jsx:
```diff
@@ -78,3 +45,5 @@ export default class JetpackConnectionErrors extends React.Component {
 		return Object.values( errorsToDisplay ).map( error => this.renderOne( error ) );
```
TIL that `React.Fragment` objects require `key` attributes when in an array.
Thanks Kim 🙂

sergeymitr commented on packages/connection/src/class-rest-connector.php:
```diff
@@ -91,18 +91,35 @@ public function __construct( Manager $connection ) {
+						'required' => true,
+					),
+				),
+				'permission_callback' => '__return_true',
```
Callback added, thanks for pointing that out.

sergeymitr commented on _inc/client/components/jetpack-notices/jetpack-connection-errors.jsx:
```diff
@@ -78,3 +45,5 @@ export default class JetpackConnectionErrors extends React.Component {
 	}
 }
+
+export default connect( null, null )( JetpackConnectionErrors );
```
It's probably a leftover, good catch.
Removed.

sergeymitr committed:
Reconnect Notice: Adding the capability to the `connection/reconnection` endpoint unit test.

kbrown9 reviewed (approved):
The changes look good to me!

I tested with both the iframe and Calypso reconnect flows, and everything worked well. I also sent an `action: reconnect` POST request to the reconnect endpoint, and the request returned with a 401 status and an invalid user permission message, as expected.
